### PR TITLE
[gas] introduce new mechanism to charge storage fees based on fixed APT prices & per-category gas/fee limits

### PR DESCRIPTION
**Pull request by vgao1996**
This makes a few additions to the gas metering system:
1. This introduces a new mechanism to charge storage-related fees based on flat APT prices that doesn't fluctuate according to the gas fee market, which would eventually allow us to decouple the execution costs and the storage costs and lower the execution costs greatly. 
2. This adds a few per-category gas/fee limits, which would allow us to increase the total amount of gas allowed without necessarily resulting in long execution time.

Note that I'm still waiting for @msmouse to come up with the correct prices. Should also mention that it might look like we're double charging (intrinsic, write set gas for io v.s. storage fees) but that is not the case. Going forward, the former group will cover transient costs only, leaving the more permanent portion to the latter. We will adjust the gas parameters to make sure we make transactions significantly cheaper, NOT more expensive.

Some of the storage fees will be replaced by deposits in the future: https://github.com/aptos-labs/aptos-core/pull/6514

### CODE REVIEW

File: aptos-move/e2e-tests/goldens/language_e2e_testsuite__tests__verify_txn__test_open_publishing.exp [hunk 1]
```diff
@@ -64,13 +64,13 @@ Ok(
...
                                 hash: OnceCell(Uninit),
-                            }: Modification(75def505000000000100000000000000),
+                            }: Modification(661af505000000000100000000000000),
                         },
                     },
                 ),
             ),
             events: [],
-            gas_used: 650,
+            gas_used: 50841,
```
Comment (vgao1996) [inline aptos-move/e2e-tests/goldens/language_e2e_testsuite__tests__verify_txn__test_open_publishing.exp:73]: This is expected because we set the gas unit price to 1 in this test. Shall we change it to 100 so that looks more similar to the range of numbers we had?

File: aptos-move/aptos-gas/src/transaction/mod.rs [hunk 2]
```diff
@@ -138,6 +165,59 @@ impl TransactionGasParameters {
...
+                Some(excess) => excess * self.storage_fee_per_excess_state_byte,
+                None => 0.into(),
+            }
+        };
+
+        let mut fee = Fee::zero();
+        let mut new_slots = NumSlots::zero();
+
+        for (key, op) in ops {
+            if let WriteOp::Creation(..) = op {}
```
Comment (grao1991) [inline aptos-move/aptos-gas/src/transaction/mod.rs:184]: Why do we have an empty if here?

Comment (vgao1996) [reply aptos-move/aptos-gas/src/transaction/mod.rs]: Good catch. Forgot to get this deleted. It's now fixed with an update.

File: types/src/event.rs [hunk 3]
```diff
@@ -45,6 +45,11 @@ impl EventKey {
         let salt = rng.next_u64();
         EventKey::new(salt, AccountAddress::random())
     }
+
+    pub fn size(&self) -> usize {
+        8 /* u64 */ + 32 /* address */
```
Comment (msmouse) [inline types/src/event.rs:50]: I'd do `std::mem::size_of<Self>()`..

**Comment by msmouse** [comment]
> The storage fees will be replaced by deposits in the future: https://github.com/aptos-labs/aptos-core/pull/6514

Not exactly :-P 
The `storage_fee_per_slot_create` fee will be collected to the global storage deposit and subject to refund, but other fees will still go with other parts of the gas and be burned or redistributed as we see proper.

File: types/src/event.rs (hunk 3 above)
Comment (grao1991) [reply types/src/event.rs:50]: Does this produce same result when different alignments are used?

Comment (vgao1996) [reply types/src/event.rs:50]: Yeah I also wanted to use `mem::size_of` but then realized it would give you a false impression that this is connected to the memory footprint, which it is totally not.

**Comment by vgao1996** [comment]
@msmouse man you are dead serious on this. I've updated the description so say only some of the fees will be replaced by deposits.

File: aptos-move/aptos-gas/src/gas_meter.rs [hunk 4]
```diff
@@ -773,11 +775,65 @@ impl AptosGasMeter {
...
+                n / d + 1
+            }
+        }
+        let gas_consumed_internal = div_ceil(
+            (u64::from(fee) as u128) * (u64::from(txn_params.gas_unit_scaling_factor) as u128),
+            u64::from(gas_unit_price) as u128,
+        );
+        let gas_consumed_internal = InternalGas::new(
+            if gas_consumed_internal > u64::MAX as u128 {
+                u64::MAX
```
[the remark this answers is not in the document]
Comment (vgao1996) [reply aptos-move/aptos-gas/src/gas_meter.rs:830]: Unless we put in some really crazy numbers in the gas schedule, this should not happen at all...

I can try to put in an `error!`, but I bet we'll run into some nasty cyclic dependency issues so I won't push it hard this time.

File: types/src/event.rs (hunk 3 above)
Comment (vgao1996) [reply types/src/event.rs:50]: I do hope there is some sort of dummy trait/functions that lets you do `bcs::serialized_size::<u64>()`

File: aptos-move/aptos-gas/src/transaction/mod.rs [hunk 5]
```diff
@@ -138,6 +165,57 @@ impl TransactionGasParameters {
...
+            match size.checked_sub(self.free_write_bytes_quota) {
+                Some(excess) => excess * self.storage_fee_per_excess_state_byte,
+                None => 0.into(),
+            }
+        };
+
+        let mut fee = Fee::zero();
+        let mut new_slots = NumSlots::zero();
+
+        for (key, op) in ops {
```
[the remark this answers is not in the document]
Comment (vgao1996) [reply aptos-move/aptos-gas/src/transaction/mod.rs:199]: Tried. That didn't work well.
1. `charge_write_set_gas` and `charge_storage_fee` use different contexts (`StorageGas` vs `TransactionGasParameters`), so combining them would require you to create a function that takes in both, which seems unnecessarily complicated.
2. It would also make the versioning logic more messy.

**Review by grao1991 — APPROVED** [review]

File: aptos-move/aptos-vm/src/aptos_vm.rs [hunk 6]
```diff
@@ -265,7 +265,13 @@ impl AptosVM {
             .finish(&mut (), gas_meter.change_set_configs())
             .map_err(|e| e.into_vm_status())?;
         // Charge gas for write set
-        gas_meter.charge_write_set_gas(user_txn_change_set_ext.write_set().iter())?;
+        gas_meter.charge_write_set_gas_for_io(user_txn_change_set_ext.write_set().iter())?;
+        gas_meter.charge_storage_fee(
```
Comment (movekevin) [inline aptos-move/aptos-vm/src/aptos_vm.rs:269]: This needs to be gated

File: aptos-move/aptos-gas/src/gas_meter.rs [hunk 7]
```diff
@@ -257,6 +266,32 @@ impl AptosGasMeter {
...
+            Ok(())
+        }
+    }
+
+    #[inline]
+    fn charge_io(&mut self, amount: InternalGas) -> PartialVMResult<()> {
+        self.charge(amount)?;
+
+        self.io_gas_used += amount;
+        if self.feature_version >= 7 && self.io_gas_used > self.gas_params.txn.max_io_gas {
```
Comment (movekevin) [inline aptos-move/aptos-gas/src/gas_meter.rs:288]: Can this break upgradability of aptos-framework package? We already have to add a mechanism to bypass tx size limit. Now that there's an io gas limit, how much headroom do we have before we need to bypass that limit as well?

File: aptos-move/aptos-gas/src/gas_meter.rs [hunk 8]
```diff
@@ -257,6 +266,32 @@ impl AptosGasMeter {
         }
     }
 
+    #[inline]
+    fn charge_execution(&mut self, amount: InternalGas) -> PartialVMResult<()> {
+        self.charge(amount)?;
+
+        self.execution_gas_used += amount;
+        if self.feature_version >= 7
+            && self.execution_gas_used > self.gas_params.txn.max_execution_gas
```
Comment (movekevin) [inline aptos-move/aptos-gas/src/gas_meter.rs:275]: I don't see max_execution_gas or max_io_gas added to the gas schedule. How are they configured?

File: types/src/event.rs [hunk 9]
```diff
@@ -45,6 +45,11 @@ impl EventKey {
         let salt = rng.next_u64();
         EventKey::new(salt, AccountAddress::random())
     }
+
+    pub fn size(&self) -> usize {
```
Comment (movekevin) [inline types/src/event.rs:49]: What is this for?

File: aptos-move/aptos-gas/src/transaction/mod.rs [hunk 10]
```diff
@@ -125,6 +133,41 @@ crate::params::define_gas_parameters!(
             { 5.. => "max_bytes_all_events_per_transaction"},
             10 << 20, // all events from a single transaction are 10MB max
         ],
+        [
+            storage_fee_per_state_slot_create: FeePerSlot,
+            { 7.. => "storage_fee_per_state_slot_create" },
+            50000,
```
Comment (movekevin) [inline aptos-move/aptos-gas/src/transaction/mod.rs:139]: How did you come up with these numbers?

**Review by movekevin — COMMENTED** [review]
Can we do benchmarking here? I'd also suggest a replay/analysis of past testnet/mainnet txs to understand how many txs would hit these limit.

File: aptos-move/aptos-vm/src/aptos_vm.rs (hunk 6 above)
Comment (vgao1996) [reply aptos-move/aptos-vm/src/aptos_vm.rs:269]: It's gated inside `charge_write_set_gas_for_io` and `charge_storage_fee`.

File: aptos-move/aptos-gas/src/transaction/mod.rs (hunk 10 above)
Comment (vgao1996) [reply aptos-move/aptos-gas/src/transaction/mod.rs:139]: @msmouse has a doc that explains the logic.

File: aptos-move/aptos-gas/src/gas_meter.rs (hunk 7 above)
Comment (vgao1996) [reply aptos-move/aptos-gas/src/gas_meter.rs:288]: This is a great question but conceptually, I don't think this is going to break framework upgrades. Our goal here is to reduce transaction costs, not increasing them. Having separate limits for each individual limits will likely result in more transactions being allowed -- old transactions should be able to pass as long as we configure the new limits no stricter than the old one.

**Comment by vgao1996** [comment]
@movekevin We are still discussing the actual gas parameter changes and the values you see in this PR are tentative. Let's perform the analysis once we get them finalized.

File: aptos-move/aptos-gas/src/transaction/mod.rs (hunk 10 above)
Comment (clay-aptos) [reply aptos-move/aptos-gas/src/transaction/mod.rs:139]: This appears to be a significant change. How should we update our docs to reflect it? Please let me help. Thanks!

Comment (vgao1996) [reply aptos-move/aptos-gas/src/transaction/mod.rs:139]: @clay-aptos thank you so much for helping! We're still in the process of figuring out the actual value changes, but I anticipate the doc work to cover the following:
1. Explaining the new system to our users, especially how the flat-APT prices are converted back to gas units to maintain compatibility
2. Making it clear that this is part of the effort to reduce cost (certainly not increasing it)
3. What would come in the future (deposits)

File: aptos-move/aptos-gas/src/aptos_framework.rs [hunk 11]
```diff
@@ -104,7 +104,7 @@ crate::natives::define_gas_parameters_for_natives!(GasParameters, "aptos_framewo
     [.code.request_publish.per_byte, "code.request_publish.per_byte", 2 * MUL],
 
     // Note(Gas): These are storage operations so the values should not be multiplied.
-    [.event.write_to_event_store.base, "event.write_to_event_store.base", 500_000],
+    [.event.write_to_event_store.base, "event.write_to_event_store.base", 300_000],
```
Comment (zekun000) [inline aptos-move/aptos-gas/src/aptos_framework.rs:107]: is this intended?

File: crates/transaction-emitter-lib/src/emitter/mod.rs [hunk 12]
```diff
@@ -41,7 +41,7 @@ use std::{
 use tokio::{runtime::Handle, task::JoinHandle, time};
 
 // Max is 100k TPS for a full day.
-const MAX_TXNS: u64 = 10_000_000_000;
+const MAX_TXNS: u64 = 100_000_000;
```
Comment (zekun000) [inline crates/transaction-emitter-lib/src/emitter/mod.rs:44]: this seems unnecessary change after updating the MAX_GAS_AMOUNT?

File: aptos-move/aptos-gas/src/transaction/mod.rs [hunk 13]
```diff
@@ -45,7 +53,7 @@ crate::params::define_gas_parameters!(
         [
             maximum_number_of_gas_units: Gas,
             "maximum_number_of_gas_units",
-            2_000_000
+            200_000_000
```
Comment (davidiw) [inline aptos-move/aptos-gas/src/transaction/mod.rs:56]: why did we increase this? where is the maximum execution units? I thought we were focusing on max execution units and then max gas units would be the additional available for storage?

**Review by davidiw — COMMENTED** [review]
can we do a quick cleanup of the numbers? so that we're roughly back at the same state we were before?

File: aptos-move/aptos-gas/src/transaction/mod.rs (hunk 13 above)
Comment (vgao1996) [reply aptos-move/aptos-gas/src/transaction/mod.rs:56]: My bad, it looks like I forgot to replace this with MAX_GAS_AMOUNT, which is now gated by a cargo feature flag.

File: crates/transaction-emitter-lib/src/emitter/mod.rs (hunk 12 above)
Comment (vgao1996) [reply crates/transaction-emitter-lib/src/emitter/mod.rs]: Reverted.

File: aptos-move/aptos-gas/src/aptos_framework.rs (hunk 11 above)
Comment (vgao1996) [reply aptos-move/aptos-gas/src/aptos_framework.rs:107]: Yes, this is set according to Alden's doc. Although it should be clear that we still haven't finalized those values and they are subject to change

**Review by zekun000 — APPROVED** [review]
awesome work!

File: aptos-move/aptos-gas/src/transaction/mod.rs (hunk 10 above)
Comment (clay-aptos) [reply aptos-move/aptos-gas/src/transaction/mod.rs:139]: Adding the *release-notes* tag so we consider highlighting this.
